tapis-project/authenticator
# fix: layout & text of authorize & login forms

## Overview

The "TTL" (Time To Live) value input for the Authorize form was too large and butting up against "Approve" checkbox. This fixes that, but reveals a Core-Styles bug, so updates to patch to fix that.

> **Important**
> There are two designs for this. I review with designers and will request leadership approval.

## Related

- closes #48
- edited by https://github.com/wesleyboar/tapis-project-authenticator/pull/1

## Changes

- **removes** extra space
- **adds** tag around variables
- **fixes** fields that were combined but should not have been

## Testing

1. Login such that you will be asked to approve token.
2. Verify checkbox field layout is not bumping against "Approve" checkbox.
3. Verify default value for TTL is shown as placeholder.

> **Warning**
> My testing has been _in situ_ on dev server.
>
> <details><summary>HTML (for Design A)</summary>
>
> ```html
> <body>
>   <style>
>     #auth-form var {
>       font-style: normal;
>       /* overwrite browser */
>     }
> 
>     /* To layout form elements in a way that lets us re-arrange them */
>     #auth-form {
>       display: flex;
>       flex-wrap: wrap;
>       justify-content: center;
>     }
> 
>     /* FAQ: If TTL field is shown, then it should be on same line as button */
>     #auth-form> :where(#ttl-field, #ttl-field + #button-container) {
>       width: 50%;
>       margin-block: 0;
>       /* to undo Core-Styles */
>     }
> 
>     /* FAQ: Other form elements should be full width */
>     #auth-form> :not(#ttl-field, #ttl-field + #button-container) {
>       width: 100%;
>     }
> 
>     /* FAQ: To vertically center TTL field (because button makes the row taller) */
>     #ttl-field {
>       justify-content: center;
>     }
> 
>     #ttl-field label {
>       margin-bottom: 0;
>     }
> 
>     /* FAQ: If TTL field is hidden, then reduce space between button and desc. */
>     :not(#ttl-field)+#button-container {
>       justify-content: center;
>       /* to override Core-Styles */
>       margin-top: 0;
>       /* to override Core-Styles */
>     }
> 
>     :not(#ttl-field)+#button-container>button {
>       margin-left: unset;
>       /* to undo Core-Styles */
>     }
> 
>     /* To remove excess space from margins that are not collapsing */
>     /* FAQ: The `#auth-form` has `dispaly: flex` which makes margins NOT collapse */
>     #auth-form h1 {
>       margin-bottom: 0;
>       /* override Core-Styles */
>     }
>   </style>
>   <main class="s-form-page">
>     <form action="" method="post" id="auth-form" class="s-form s-form--login">
>       <h1>
>         <figure>
>           <img src="authorize/tapis.png" alt="Tapis">
>           <img src="authorize/tacc-formal.svg" alt="TACC">
>         </figure><span>Connect to Tapis</span>
>       </h1>
>       <p>The <var>Portal Name Here Portal</var> wants to use Tapis (TACC APIs) service to enable actions on behalf of
>         <var>ab12345@utexas.edu</var>.</p>
>       <div id="ttl-field">
>         <label aria-labelledby="ttl-label"> Allow this authorization for <input type="text" id="ttl" name="ttl"
>             value="30" size="3"> days. </label>
>         <span hidden="" id="ttl-label">Allow this authorization for how many days? By default, the authorization will
>           last 30 days. </span>
>       </div>
>       <footer id="button-container">
>         <button type="submit">Connect</button>
>       </footer>
>     </form>
>   </main>
>   <footer class="s-footer">
>     <p>Tapis project funded by NSF OAC grants <a href="https://www.nsf.gov/awardsearch/showAward?AWD_ID=1931439"
>         rel="noreferrer" target="_blank">#1931439</a>, <a
>         href="https://www.nsf.gov/awardsearch/showAward?AWD_ID=1931575" rel="noreferrer" target="_blank">#1931575</a>
>       and #2232259. </p>
>   </footer>
> </body>
> ```
>
> </details>


## UI

There are two designs:

| | pro & con | code |
| - | - | - |
| A | clear, but [requires WMA](https://tacc-team.slack.com/archives/C04TQ1P04UX/p1694727181558209?thread_ts=1694726906.175339&cid=C04TQ1P04UX) set `client_display_name` | #49 (this) |
| B | requires no WMA effort, but not clear to user | [wesleyboar#1](https://github.com/wesleyboar/tapis-project-authenticator/pull/1) |

| | sans TTL field | with TTL field |
| - | - | - |
| A | ![Connect](https://github.com/tapis-project/authenticator/assets/62723358/8a6affa9-7854-4c60-a555-e403b90a0d4d) | ![Connect (with TTL)](https://github.com/tapis-project/authenticator/assets/62723358/36ccddcf-fefd-4e44-b210-6570ba3c57dc)
| B | ![Authorize](https://github.com/tapis-project/authenticator/assets/62723358/3084d51a-38c2-4c37-b6b3-42bdc14b7fa1) | ![Authorize (with TTL)](https://github.com/tapis-project/authenticator/assets/62723358/c3c4164e-d82f-4713-b6dd-09674cdc253a)

| accessibility |
| - |
| ![Auth - TTL A11y](https://github.com/tapis-project/authenticator/assets/62723358/896a66fb-ecb4-40be-8e63-3f4796227fb3) |

<details><summary>Archive</summary>

### As of 75c53a4

| with TTL field | sans TTL field |
| - | - |
| ![with ttl field](https://github.com/tapis-project/authenticator/assets/62723358/9939100d-f255-4707-ada7-5533562dacab) | ![sans ttl field](https://github.com/tapis-project/authenticator/assets/62723358/637911c8-7362-4042-9ccf-e8bdd2f7d02f) |

| accessibility |
| - |
| ![a11y](https://github.com/tapis-project/authenticator/assets/62723358/024f02cc-7d17-4a2c-96c0-fc364a1f2538) |

### As of 562054b

| before (in situ test) | after (in situ test) |
| - | - |
| <img width="660" alt="before" src="https://github.com/tapis-project/authenticator/assets/62723358/6e3bfa85-fda2-4c6f-95f0-4bdc8f3012ce"> | <img width="660" alt="after" src="https://github.com/tapis-project/authenticator/assets/62723358/d5c8b117-475e-40d2-beae-644e8f6bc5e3"> |

| accessibility |
| - |
|  <img width="660" alt="a11y" src="https://github.com/tapis-project/authenticator/assets/62723358/76c1cb81-a2e8-4042-883c-cb464cedf2ee"> |

</details>
